frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx

```diff
@@ -0,0 +1,332 @@
+      uploadTheFile(file)
+
+      ws.setOptions({
+        // This color is hardcoded because I've spent an hour trying to get it to work
+        // with the theme and I'm giving up for now. The complications arise due to the
+        // way wavesurfer is using these colors. Specifically, the progress color "tints" the wave color
+        // and the opacity of this wave color is not being respected, so to make things work for now hardcoding
+        // this color that looks "okay" on both themes.
+        waveColor: "#A5A5AA",
+      })
+    })
+
+    recordPlugin.on("record-progress", time => {
+      setRecordingTime(formatTime(time))
+    })
+
+    setWavesurfer(ws)
+    setRecordPlugin(recordPlugin)
+
+    return () => {
+      if (ws) ws.destroy()
+    }
+  }, [theme, uploadTheFile])
+
+  useEffect(() => {
+    initializeWaveSurfer()
+  }, [initializeWaveSurfer])
+
+  const onClickPlayPause = useCallback(() => {
+    if (wavesurfer) {
+      wavesurfer.playPause()
+      // This is because we want the time to be the duration of the audio when they stop recording,
+      // but once they start playing it, we want it to be the current time. So, once they start playing it
+      // we'll start keeping track of the playback time from that point onwards (until re-recording).
+      setShouldUpdatePlaybackTime(true)
+
+      // to get the pause button to show
+      forceRerender()
+    }
+  }, [wavesurfer])
+
+  const startRecording = useCallback(() => {
+    if (!recordPlugin || !activeAudioDeviceId || !wavesurfer) {
+      return
+    }
+
+    const deviceId = activeAudioDeviceId
+    if (isNullOrUndefined(deviceId)) {
+      return
+    }
+
+    wavesurfer.setOptions({
+      waveColor: theme.colors.primary,
+    })
+
+    recordPlugin.startRecording({ deviceId: activeAudioDeviceId }).then(() => {
+      // Update the record button to show the user that they can stop recording
+      forceRerender()
+    })
+  }, [activeAudioDeviceId, recordPlugin, theme, wavesurfer])
+
+  const stopRecording = useCallback(() => {
+    if (!recordPlugin) return
+
+    recordPlugin.stopRecording()
+  }, [recordPlugin])
+
+  const handleClear = useCallback(() => {
+    if (isNullOrUndefined(wavesurfer) || isNullOrUndefined(deleteFileUrl)) {
+      return
+    }
+    setRecordingUrl(null)
+    wavesurfer.empty()
+    uploadClient.deleteFile(deleteFileUrl)
+    setProgressTime(STARTING_TIME_STRING)
+    setRecordingTime(STARTING_TIME_STRING)
+    setDeleteFileUrl(null)
+    widgetMgr.setFileUploaderStateValue(
+      element,
+      {},
+      { fromUi: true },
+      undefined
+    )
+    setShouldUpdatePlaybackTime(false)
+    if (notNullOrUndefined(recordingUrl)) {
+      URL.revokeObjectURL(recordingUrl)
+    }
+  }, [
+    deleteFileUrl,
+    recordingUrl,
+    uploadClient,
+    wavesurfer,
+    element,
+    widgetMgr,
+  ])
+
+  // Note: these can't be memoized due to the reliance on calling .isRecording() & .isPlaying()
```

Comment:
though you could store `const isRecording = recordPlugin.isRecording()` here and use it in the next two checks if you believe that it is an expensive call